libcudacxx/include/cuda/std/detail/libcxx/include/__cuda/barrier.h

```diff
@@ -641,176 +678,423 @@ _LIBCUDACXX_INLINE_VISIBILITY constexpr bool __unused(_Ty...) {return true;}
-                        : "memory");
-                }
-            )
-        )
-        return async_contract_fulfillment::async;
+    template<typename _Group, thread_scope _Sco, typename _CompF>
+    _LIBCUDACXX_INLINE_VISIBILITY static
+    async_contract_fulfillment __delay(
+        __completion_mechanism __cm, _Group const & __group, _CUDA_VSTD::size_t __size, barrier<_Sco, _CompF> & __barrier) {
+        return __delay_non_smem_barrier(__cm, __group, __size, __barrier);
     }
-};
 
-template<>
-struct __memcpy_async_impl<16, false> {
-    _LIBCUDACXX_DEVICE static inline async_contract_fulfillment __copy(char * __destination, char const * __source, _CUDA_VSTD::size_t __total_size, _CUDA_VSTD::size_t __rank, _CUDA_VSTD::size_t __stride) {
-        // Guard from host visibility when compiling in host only contexts
-        NV_IF_TARGET(
-            NV_IS_DEVICE, (
-                for (_CUDA_VSTD::size_t __offset = __rank * 16; __offset < __total_size; __offset += __stride * 16) {
-                    asm volatile ("cp.async.cg.shared.global [%0], [%1], 16, 16;"
-                        :: "r"(static_cast<_CUDA_VSTD::uint32_t>(__cvta_generic_to_shared(__destination + __offset))),
-                            "l"(__source + __offset)
-                        : "memory");
-                }
-            )
-        )
+    template<typename _Group, thread_scope _Sco, typename _CompF>
+    _LIBCUDACXX_INLINE_VISIBILITY static
+    async_contract_fulfillment __delay_non_smem_barrier(
+        __completion_mechanism __cm, _Group const & __group, _CUDA_VSTD::size_t __size, barrier<_Sco, _CompF> & __barrier) {
+        // Overload for non-smem barriers.
+        if (__cm == __completion_mechanism::__async_group) {
+            // Pre-SM80, the async_group mechanism is not available.
+            NV_IF_TARGET(NV_PROVIDES_SM_80, (
+                // Blocking: wait for all thread-local cp.async instructions to have
+                // completed writing to shared memory.
+                asm volatile ("cp.async.wait_all;" ::: "memory");
+            ));
         return async_contract_fulfillment::async;
+        } else if (__cm == __completion_mechanism::__mbarrier_complete_tx) {
+            // Non-smem barriers do not have an mbarrier_complete_tx mechanism..
+            // XXX: should be __builtin_unreachable() or an assert?
+            return async_contract_fulfillment::none;
+        } else {
+            // sync: In this case, we do not need to do anything.
+            return async_contract_fulfillment::none;
+        }
+    }
+
+    template<typename _Group, thread_scope _Sco>
+    _LIBCUDACXX_INLINE_VISIBILITY static
+    async_contract_fulfillment __delay(__completion_mechanism __cm, _Group const & __group, _CUDA_VSTD::size_t __size, pipeline<_Sco> & __pipeline) {
+        // pipeline does not sync on memcpy_async, defeat pipeline purpose otherwise
+        __unused(__pipeline);
+        __unused(__size);
+        __unused(__group);
```

Comment:
```suggestion
    async_contract_fulfillment __delay(__completion_mechanism __cm, _Group const&, _CUDA_VSTD::size_t, pipeline<_Sco>&) {
        // pipeline does not sync on memcpy_async, defeat pipeline purpose otherwise
```